### PR TITLE
Refactor category orders in heads

### PR DESCRIPTION
**Pull request by ZwwWayne**
This PR changes the orders of categories in both anchor and box heads.
The background is no longer regarded as the first class anymore. Rather, it is taken as the last category in heads.

### CODE REVIEW

**Comment by ZwwWayne** [comment]
For now, Mask RCNN, Retina_free_anchor, ATSS, MS RCNN, HTC, ga_faster, ga_retinanet,FCOS, FoveaBox, PepPoints, and RetinaNet have been proven to be trained successfully. The performance will come out soon.

File: mmdet/models/anchor_heads/anchor_head.py [hunk 1]
```diff
@@ -16,7 +16,7 @@ class AnchorHead(nn.Module):
     """Anchor-based head (RPN, RetinaNet, SSD, etc.).
 
     Args:
-        num_classes (int): Number of categories including the background
+        num_classes (int): Number of categories not including the background
```
Comment (hellock) [inline mmdet/models/anchor_heads/anchor_head.py:19]: not including -> excluding

File: .gitignore [hunk 2]
```diff
@@ -117,3 +117,4 @@ work_dirs/
 # Pytorch
 *.pth
 *.py~
+./exps/*
```
Comment (xvjiarui) [inline .gitignore:120]: Is that necessary for our codebase?

File: mmdet/core/anchor/anchor_target.py [hunk 3]
```diff
@@ -130,13 +134,17 @@ def anchor_target_single(flat_anchors,
...
+    # remind new system set FG cat_id: [0, num_class-1], BG cat_id: num_class
+    if gt_labels is not None:
+        labels += num_classes
     if len(pos_inds) > 0:
         pos_bbox_targets = bbox2delta(sampling_result.pos_bboxes,
                                       sampling_result.pos_gt_bboxes,
                                       target_means, target_stds)
         bbox_targets[pos_inds, :] = pos_bbox_targets
         bbox_weights[pos_inds, :] = 1.0
         if gt_labels is None:
```
Comment (xvjiarui) [inline mmdet/core/anchor/anchor_target.py:146]: Could we deal with rpn condition together to reduce ambiguity?

File: mmdet/models/losses/ghm_loss.py [hunk 4]
```diff
@@ -7,9 +7,9 @@
 
 def _expand_binary_labels(labels, label_weights, label_channels):
```
Comment (xvjiarui) [inline mmdet/models/losses/ghm_loss.py:8]: If the label is not binary, should we rename this function?

File: mmdet/core/anchor/anchor_target.py [hunk 5]
```diff
@@ -26,6 +27,7 @@ def anchor_target(anchor_list,
         target_means (Iterable): Mean value of regression targets.
         target_stds (Iterable): Std value of regression targets.
         cfg (dict): RPN train configs.
+        background_label (int): label id of background
```
Comment (hellock) [inline mmdet/core/anchor/anchor_target.py:30]: The arg list does not match the actual one.

File: mmdet/core/evaluation/mean_ap.py [hunk 6]
```diff
@@ -429,7 +430,7 @@ def print_map_summary(mean_ap,
         num_gts[:, i] = cls_result['num_gts']
 
     if dataset is None:
-        label_names = [str(i) for i in range(1, num_classes + 1)]
+        label_names = [str(i) for i in range(0, num_classes)]
```
Comment (hellock) [inline mmdet/core/evaluation/mean_ap.py:433]: `range(num_classes)`

File: mmdet/core/anchor/anchor_target.py [hunk 7]
```diff
@@ -125,7 +136,8 @@ def anchor_target_single(flat_anchors,
     num_valid_anchors = anchors.shape[0]
     bbox_targets = torch.zeros_like(anchors)
     bbox_weights = torch.zeros_like(anchors)
-    labels = anchors.new_zeros(num_valid_anchors, dtype=torch.long)
+    labels = anchors.new_zeros(
+        num_valid_anchors, dtype=torch.long) + background_label
```
Comment (xvjiarui) [inline mmdet/core/anchor/anchor_target.py:140]: We could change it to `labels = anchors.new_empty(num_valid_anchors, dtype=torch.long).fill_(background_label)`.
Just to be consistent with `bbox_target`

File: mmdet/models/anchor_heads/atss_head.py [hunk 8]
```diff
@@ -444,7 +452,8 @@ def atss_target_single(self,
         num_valid_anchors = anchors.shape[0]
         bbox_targets = torch.zeros_like(anchors)
         bbox_weights = torch.zeros_like(anchors)
-        labels = anchors.new_zeros(num_valid_anchors, dtype=torch.long)
+        labels = anchors.new_zeros(
+            num_valid_anchors, dtype=torch.long) + background_label
```
Comment (xvjiarui) [inline mmdet/models/anchor_heads/atss_head.py:456]: We could change it to `labels = anchors.new_empty(num_valid_anchors, dtype=torch.long).fill_(background_label)`.

File: mmdet/core/anchor/anchor_target.py (hunk 7 above)
Comment (ZwwWayne) [reply mmdet/core/anchor/anchor_target.py:140]: sure

File: mmdet/core/mask/structures.py [hunk 9]
```diff
@@ -90,6 +90,11 @@ def __getitem__(self, index):
     def __iter__(self):
         return iter(self.masks)
 
+    def __repr__(self) -> str:
+        s = self.__class__.__name__ + '('
+        s += 'num_masks={})'.format(len(self.masks))
```
Comment (hellock) [inline mmdet/core/mask/structures.py:95]: Include the height and width as well.

File: mmdet/models/anchor_heads/free_anchor_retina_head.py [hunk 10]
```diff
@@ -66,7 +66,7 @@ def loss(self,
         for _, (anchors_, gt_labels_, gt_bboxes_, cls_prob_,
                 bbox_preds_) in enumerate(
                     zip(anchors, gt_labels, gt_bboxes, cls_prob, bbox_preds)):
-            gt_labels_ -= 1
+            # gt_labels_ -= 1
```
Comment (hellock) [inline mmdet/models/anchor_heads/free_anchor_retina_head.py:69]: Remove the comment?

File: mmdet/models/roi_heads/cascade_roi_head.py [hunk 11]
```diff
@@ -163,6 +163,10 @@ def _mask_forward_train(self,
                             rcnn_train_cfg,
                             bbox_feats=None):
         pos_rois = bbox2roi([res.pos_bboxes for res in sampling_results])
+        if len(pos_rois) == 0:
+            # If there are no predicted and/or truth boxes, then we cannot
+            # compute head / mask losses
+            return dict(loss_mask=None)
```
Comment (hellock) [inline mmdet/models/roi_heads/cascade_roi_head.py:169]: Is this case tested?

File: mmdet/core/mask/structures.py [hunk 12]
```diff
@@ -90,6 +90,11 @@ def __getitem__(self, index):
     def __iter__(self):
         return iter(self.masks)
 
+    def __repr__(self) -> str:
```
Comment (hellock) [inline mmdet/core/mask/structures.py:93]: Add type hints in a future PR.
